cmd/clusterctl/pkg/client/cluster/components_test.go

```diff
@@ -0,0 +1,221 @@
+			},
+			wantErr: false,
+		},
+
+		{
+			name: "Delete provider, provider namespace and provider CRDs",
+			args: args{
+				provider:             clusterctlv1.Provider{ObjectMeta: metav1.ObjectMeta{Name: "aws", Namespace: "ns1"}},
+				forceDeleteNamespace: true,
+				forceDeleteCRD:       true,
+			},
+			wantDiff: []wantDiff{
+				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Namespace", Name: "ns1"}, deleted: true},                                           //namespace should be deleted
+				{object: corev1.ObjectReference{APIVersion: "apiextensions.k8s.io/v1beta1", Kind: "CustomResourceDefinition", Name: "crd1"}, deleted: true}, //crd should be deleted
+				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: "ns1", Name: "pod1"}, deleted: true},                              // provider components should be deleted
+				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: "ns1", Name: "pod2"}, deleted: true},                              // other objects in the namespace goes away when deleting the namespace
+				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: "ns2", Name: "pod3"}, deleted: false},                             // this object is in another namespace, and should never be touched by delete
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			proxy := test.NewFakeProxy().WithObjs(initObjs...)
+			c := newComponentsClient(proxy)
+			err := c.Delete(DeleteOptions{
+				Provider:             tt.args.provider,
+				ForceDeleteNamespace: tt.args.forceDeleteNamespace,
+				ForceDeleteCRD:       tt.args.forceDeleteCRD,
+			})
+			if (err != nil) != tt.wantErr {
+				t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if tt.wantErr {
+				return
+			}
+
+			cs, err := proxy.NewClient()
+			if err != nil {
+				t.Fatalf("NewClient() error = %v", err)
+			}
+			for _, want := range tt.wantDiff { // Want diff use OP\\GroupVersion\Kind\Name syntax for expected object changes (OP: = not delete, - deleted)
+				obj := &unstructured.Unstructured{}
+				obj.SetAPIVersion(want.object.APIVersion)
+				obj.SetKind(want.object.Kind)
+
+				key := client.ObjectKey{
+					Namespace: want.object.Namespace,
+					Name:      want.object.Name,
+				}
+
+				err := cs.Get(ctx, key, obj)
+				if err != nil && !apierrors.IsNotFound(err) {
+					t.Fatalf("Failed to get %v from the cluster: %v", key, err)
+				}
+
+				if !want.deleted && apierrors.IsNotFound(err) {
+					t.Errorf("%v deleted, expect NOT deleted", key)
+				}
+
+				if want.deleted && !apierrors.IsNotFound(err) {
+					if want.object.Namespace == "ns1" && tt.args.forceDeleteNamespace { // Ignoring namespaced object that should be deleted by the namespace controller.
```

Comment:
Would it be slightly safer/future-proof to match want.object.Namespace with tt.args.provider.Namespace instead of hard-coding ns1?

Reply:
Done. Nice one